myia/unify.py

```diff
@@ -481,3 +482,153 @@ def reify(self, v: Value[T], equiv: EquivT[T]) -> Value[T]:
+
+class DomainDict(Dict[str, Any]):
+    """Hashable dict for domain unification."""
+
+    __slots__ = ('_hash')
+
+    def __hash__(self):
+        if not hasattr(self, '_hash'):
+            keys = tuple(sorted(list(self.keys())))
+            vals = tuple(self[k] for k in self.keys())
+            self._hash = hash((keys, vals))
+        return self._hash
+
+
+def ddict(**kwargs: Dict[str, Any]) -> DomainDict:
+    """Shorthand to create DomainDicts faster."""
+    return DomainDict(**kwargs)
+
+
+class DomainUnification(Unification[Any]):
+    """Unification that can work across different domains.
+
+    This means that you can run unification on two different sets of
+    values and sometimes reach acorss the sets for a value that you
+    need.
+
+    The domains are identified by strings and all the user-facing
+    functions require DomainDicts as input to indetify the staring
+    domain, unless you use the .domain() context manager.
+
+    You don't need to identify the domain at each level as it will
+    assume that domain of a value is the same as its parent unless it
+    encounters a new DomainDict.
+
+    This means that something like::
+
+       ddict(dom1=(val1, val2))
+
+    will assume domain 'dom1' for val1 and val2.  If you want to
+    switch Domains you can do::
+
+      ddict(dom1=(val1, ddict(dom2=val2)))
+
+    This will use 'dom1' for val1 (and the tuple) and 'dom2' for val2.
+
+    This class also changes slightlt the contract of the unify and
+    unify_raw methods in the presence of DomainDicts.  They require
+    that all the domains in the first dictionary are present in the
+    second, but not the inverse.
+
+    """
+
+    def __init__(self, domains: Dict[str, Callable]) -> None:
+        """Supply intial domains."""
+        self._domains = domains
+        self._domstack: List[Callable] = []
+
+    def add_domain(self, name: str, dom: Callable):
+        """Register a new domain.
+
+        You can add domains via this method (provided the name doesn't
+        already exist) but there is no way to remove a domain once
+        it's there.  This is because removing domain could break the
+        navigation of equivalences.
+        """
+        assert name not in self._domains
+        self._domains[name] = dom
+
+    @contextmanager
+    def domain(self, name: str):
+        """Context manager to set the "current" domain for values.
+
+        This will be overridden if a DomainDict is encountered.
+
+        """
+        self._dompush(self._domains[name])
+        try:
+            yield self
+        finally:
+            self._dompop()
+
+    def _curdom(self):
+        if len(self._domstack):
+            return self._domstack[-1]
+        return None
+
+    def _dompush(self, dom):
+        self._domstack.append(dom)
+
+    def _dompop(self):
+        self._domstack.pop(-1)
+
+    def visit(self, fn, value):
+        """Visit according to current domain."""
+        return self._curdom()(fn, value)
```

Comment:
I think I brought it up in conversation a couple weeks ago but I'm going to raise this issue again: wouldn't it be simpler to have one universal `visit` method? Support for new types could easily be added using some `register_visitor` decorator, or a `__visit__` magic method. Different domains are going to visit different types of data, but they are not going to visit the same types differently, as far as I can tell.

Reply:
Different domains have different terminators.  This is not possible with a universal visit method.

For example, when you are matching values and you encounter `Tuple(Float(64), Float(64))`, you don't need to reach inside the tuple.  If Tuple had a visitor for it, you wouldn't be able to make that distinction.

Reply:
We discussed in person and we will move to a model with global visit.